### PR TITLE
refactor: use SqlArgument wrapper to look up functions

### PR DESCRIPTION
**Pull request by stevenpyzhang**
### Description 
Uses SqlArgument wrapper when looking up functions. This allows us in future PR's to handle non SqlType function arguments (such as lambdas)

PR looks large, but it's mainly just replacing instances of SqlType with SqlArgument in UDF lookup code.

### Testing done 
Updated unit tests, non functional change

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java [hunk 1]
```diff
@@ -172,11 +173,15 @@ public static SqlType applyResolved(
...
+  // CHECKSTYLE_RULES.OFF: CyclomaticComplexity
   private static boolean resolveGenerics(
       final List<Entry<GenericType, SqlType>> mapping,
       final ParamType schema,
-      final SqlType instance
+      final SqlArgument instance
   ) {
+    final SqlType sqlType = instance.getSqlType();
+
+    // CHECKSTYLE_RULES.ON: CyclomaticComplexity
```
Comment (lct45) [inline ksqldb-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java:184]: Is there a usual pattern for where we turn the checkstyle rules back on?

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/TopKDistinctTestUtils.java [hunk 2]
```diff
@@ -16,7 +16,9 @@
 package io.confluent.ksql.function.udaf.topkdistinct;
 
 import io.confluent.ksql.function.AggregateFunctionInitArguments;
+import io.confluent.ksql.schema.ksql.SqlArgument;
 import io.confluent.ksql.schema.ksql.types.SqlType;
+
```
Comment (lct45) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/TopKDistinctTestUtils.java:21]: Does the IDE add these random extra lines?

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java [hunk 3]
```diff
@@ -432,9 +434,11 @@ private String formatExpression(final Expression expression) {
       final String instanceName = funNameToCodeName.apply(functionName);
 
       final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName());
-      final List<SqlType> argumentSchemas = node.getArguments().stream()
-          .map(expressionTypeManager::getExpressionSqlType)
-          .collect(Collectors.toList());
+      final List<SqlArgument> argumentSchemas = new ArrayList<>();
+      for (final Expression argExpr : node.getArguments()) {
+        final SqlType newSqlType = expressionTypeManager.getExpressionSqlType(argExpr);
```
Comment (lct45) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java:439]: This is my bad, I missed this when doing some updates. We should be able to leave the original implementation and just add `map(arg -> sqlArgument.of(arg.getSqlType()))`

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/function/UdtfUtil.java [hunk 4]
```diff
@@ -41,10 +42,12 @@ public static KsqlTableFunction resolveTableFunction(
 
     final List<Expression> functionArgs = functionCall.getArguments();
 
-    final List<SqlType> argTypes = functionArgs.isEmpty()
-        ? ImmutableList.of(FunctionRegistry.DEFAULT_FUNCTION_ARG_SCHEMA)
-        : functionArgs.stream().map(expressionTypeManager::getExpressionSqlType)
-            .collect(Collectors.toList());
+    final List<SqlArgument> argTypes = functionArgs.isEmpty()
+        ? ImmutableList.of(SqlArgument.of(FunctionRegistry.DEFAULT_FUNCTION_ARG_SCHEMA))
+        : new ArrayList<>();
```
Comment (lct45) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/function/UdtfUtil.java:47]: Same as above, for a minute I couldn't figure out how to get the additional `map(arg -> sqlArgument.of(arg.sqlType())` to work, but we should be able to add that to the original implementation and avoid the for loop

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java [hunk 5]
```diff
@@ -445,10 +445,13 @@ public Void visitFunctionCall(
...
 
       if (functionRegistry.isTableFunction(node.getName())) {
-        final List<SqlType> argumentTypes = node.getArguments().isEmpty()
-            ? ImmutableList.of(FunctionRegistry.DEFAULT_FUNCTION_ARG_SCHEMA)
-            : node.getArguments().stream().map(ExpressionTypeManager.this::getExpressionSqlType)
-                .collect(Collectors.toList());
+        final List<SqlArgument> argumentTypes = node.getArguments().isEmpty()
+            ? ImmutableList.of(
+                SqlArgument.of(FunctionRegistry.DEFAULT_FUNCTION_ARG_SCHEMA))
+            : new ArrayList<>();
```
Comment (lct45) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java:451]: Same as above

**Review by lct45 — APPROVED** [review]
Thanks for pulling this out @stevenpyzhang , overall LGTM with a few questions/nits

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/TopKDistinctTestUtils.java (hunk 2 above)
Comment (stevenpyzhang) [reply ksqldb-engine/src/test/java/io/confluent/ksql/function/udaf/topkdistinct/TopKDistinctTestUtils.java]: Not sure where this came from, I removed it though

File: ksqldb-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java (hunk 1 above)
Comment (stevenpyzhang) [reply ksqldb-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java]: Just following the other patterns, we usually turn it back on right after the method signature, I moved it up one line now

**Pull request merged**